zerver/migrations/0436_fix_deleteduser_email.py

```diff
@@ -0,0 +1,66 @@
+# Generated by Django 4.2 on 2023-04-17 18:25
+
+from email.headerregistry import Address
+from functools import lru_cache
+
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.core.validators import validate_email
+from django.db import migrations
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Q
+
+
+@lru_cache
+def get_fake_email_domain(apps: StateApps, realm_id: int) -> str:
+    Realm = apps.get_model("zerver", "Realm")
+    realm = Realm.objects.get(id=realm_id)
+
+    # Build out realm.host
+    subdomain = realm.string_id
+    if subdomain == "":
+        host = settings.EXTERNAL_HOST
+    else:
+        default_host = f"{subdomain}.{settings.EXTERNAL_HOST}"
+        host = settings.REALM_HOSTS.get(subdomain, default_host)
+
+    # Implement get_fake_email_domain
+    try:
+        validate_email(Address(username="bot", domain=host).addr_spec)
+        return realm.host
+    except ValidationError:
+        pass
+
+    try:
+        # Check that the fake email domain can be used to form valid email addresses.
+        validate_email(Address(username="bot", domain=settings.FAKE_EMAIL_DOMAIN).addr_spec)
+        return settings.FAKE_EMAIL_DOMAIN
+    except ValidationError:
+        raise Exception(
+            settings.FAKE_EMAIL_DOMAIN + " is not a valid domain. "
+            "Consider setting the FAKE_EMAIL_DOMAIN setting."
+        )
+
+
+def fix_invalid_emails(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    UserProfile = apps.get_model("zerver", "UserProfile")
```

Comment:
I like to always write a docstring for migration functions that explains the problem we're fixing at a high level rather than requiring one to find the commit message via `git log` or similar.